File: MomentumAndroid/src/main/java/com/mwaibanda/momentum/android/presentation/volunteer/ServicesViewModel.kt
package com.mwaibanda.momentum.android.presentation.volunteer

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mwaibanda.momentum.domain.models.Service
import com.mwaibanda.momentum.domain.models.Tab
import com.mwaibanda.momentum.domain.models.VolunteerServiceRequest
import com.mwaibanda.momentum.domain.usecase.service.ServicesUseCases
import com.mwaibanda.momentum.utils.Result
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class ServicesViewModel(
    private val servicesUseCase: ServicesUseCases
): ViewModel() {
    private val _services = MutableStateFlow(emptyList<Tab>())
    val services = _services.asStateFlow()

    private val _service: MutableStateFlow<Service?> = MutableStateFlow(null)
    val service = _service.asStateFlow()

    private val _refreshCount = MutableStateFlow(0)
    val refreshCount = _refreshCount.asStateFlow()

    fun getServices() {
        viewModelScope.launch {
            servicesUseCase.read().collectLatest {
                when(it) {
                    is Result.Data -> {
                        _services.value = (it.data ?: emptyList())
                        Log.e("GET[Data]", "Services")
                    }
                    is Result.Error -> Log.e("ServicesViewModel[getServices]", it.message ?: "")

                    is Result.Loading -> Log.e("GET[Loading]", "Services")
                }
            }
        }
    }

    fun getServiceByType(type: Tab.Type) {
        viewModelScope.launch {
            servicesUseCase.readByType(type).collectLatest {
                when(it) {
                    is Result.Data -> {
                        _service.value = it.data
                        Log.e("GET[Data]", "Service")
                    }
                    is Result.Error -> Log.e("ServicesViewModel[getServiceByType]", it.message ?: "")

                    is Result.Loading -> Log.e("GET[Loading]", "Service")
                }
            }
        }
    }

    fun postVolunteerService(request: VolunteerServiceRequest) {
        viewModelScope.launch {
            servicesUseCase.create(request).collectLatest {
                when(it) {
                    is Result.Data -> {
                        _refreshCount.value++
                        Log.e("GET[Data]", "Service")
                    }
                    is Result.Error -> Log.e("ServicesViewModel[getServiceByType]", it.message ?: "")

                    is Result.Loading -> Log.e("GET[Loading]", "Service")
                }
            }
        }
    }
}

File: MomentumiOS/MomentumiOS/Presentation/Volunteer/ServicesViewModel.swift
//
//  ServicesViewModel.swift
//  MomentumiOS
//
//  Created by Mwai Banda on 1/11/25.
//  Copyright © 2025 Momentum. All rights reserved.
//
import Foundation
import MomentumSDK
import TinyDi
import os

class ServicesViewModel: ObservableObject {
    @Inject private var servicesUseCases: ServicesUseCases
    @Published var services = [Tab]()
    @Published var service : Service? = nil
    func getServices() {
        servicesUseCases.read.execute().collect { [weak self] res in
            guard let status = res?.status else { fatalError("No Result Found") }
            switch(status) {
            case .loading:
                os_log("[getServices[Loading]]")
                break
            case .error:
                guard let message = res?.message  else { return }
                os_log("[getServices[Error]]: \(message)")
                break
            case .data:
                guard let services = res?.data as? [Tab] else { return }
                DispatchQueue.main.async {
                    self?.services = services
                }
                os_log("[getServices[Data]]")
                break
            default: break
            }
        }
    }
    
    func getServiceByType(tab: Tab, onCompletion: @escaping ([VolunteerService]) -> Void) {
        servicesUseCases.readByType.execute(type: tab.type).collect { [weak self] res in
            guard let status = res?.status else { fatalError("No Result Found") }
            switch(status) {
            case .loading:
                os_log("[getServiceByType[Loading]]")
                break
            case .error:
                guard let message = res?.message  else { return }
                os_log("[getServiceByType[Error]]: \(message)")
                break
            case .data:
                guard let service = res?.data else { return }
                DispatchQueue.main.async {
                    self?.service = service
                    onCompletion(service.services)
                }
                os_log("[getServiceByType[Data]]")
                break
            default: break
            }
        }
    }
    
    func updateVolunteerServiceDay(request: DayRequest, onCompletion: @escaping () -> Void) {
        servicesUseCases.updateDay.execute(request: request).collect { res in
            guard let status = res?.status else { fatalError("No Result Found") }
            switch(status) {
            case .loading:
                os_log("[updateVolunteerServiceDay[Loading]]")
                break
            case .error:
                guard let message = res?.message  else { return }
                os_log("[updateVolunteerServiceDay[Error]]: \(message)")
                break
            case .data:
                onCompletion()
                os_log("[updateVolunteerServiceDay[Data]]")
                break
            default: break
            }
        }
    }
    
    func postVolunteerService(request: VolunteerServiceRequest) {
        servicesUseCases.create.execute(request: request).collect { res in
            guard let status = res?.status else { fatalError("No Result Found") }
            switch(status) {
            case .loading:
                os_log("[postVolunteerService[Loading]]")
                break
            case .error:
                guard let message = res?.message  else { return }
                os_log("[postVolunteerService[Error]]: \(message)")
                break
            case .data:
                os_log("[postVolunteerService[Data]]")
                break
            default: break
            }
        }
    }
    
}
